aspnetcore/blazor/forms-validation.md

```diff
@@ -223,12 +230,383 @@ In the following example:
+
+In the client app, create a server validation component to display error messages returned by the server API (`ServerValidator.cs`):
+
+```csharp
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Components;
+using Microsoft.AspNetCore.Components.Forms;
 
-    private async Task<bool> ServerValidate(EditContext editContext)
+namespace {SOLUTION NAMESPACE}.Client
+{
+    public class ServerValidator : ComponentBase
     {
-        var serverChecksValid = ...
+        private ValidationMessageStore messageStore;
+
+        [CascadingParameter]
+        private EditContext CurrentEditContext { get; set; }
+
+        protected override void OnInitialized()
+        {
+            if (CurrentEditContext == null)
+            {
+                throw new InvalidOperationException(
+                    $"{nameof(ServerValidator)} requires a cascading " +
+                    $"parameter of type {nameof(EditContext)}. " +
+                    $"For example, you can use {nameof(ServerValidator)} " +
+                    $"inside an {nameof(EditForm)}.");
+            }
+
+            messageStore = new ValidationMessageStore(CurrentEditContext);
+
+            CurrentEditContext.OnValidationRequested += (s, e) => 
+                messageStore.Clear();
+            CurrentEditContext.OnFieldChanged += (s, e) => 
+                messageStore.Clear(e.FieldIdentifier);
+        }
+
+        public void DisplayErrors(Dictionary<string, List<string>> errors)
+        {
+            foreach (var err in errors)
+            {
+                messageStore.Add(CurrentEditContext.Field(err.Key), err.Value);
+            }
 
-        return serverChecksValid;
+            CurrentEditContext.NotifyValidationStateChanged();
+        }
+    }
+}
+```
+
+In the client project, the *Starfleet Starship Database* form is updated to show server validation errors with the `ServerValidaor` component. When the server API returns validation messages, they're added to the form's <xref:Microsoft.AspNetCore.Components.Forms.ValidationMessageStore>:
+
+```csharp
+@page "/FormsValidation"
+@using System.Net.Http.Json
+@using Microsoft.AspNetCore.Authorization
+@using Microsoft.AspNetCore.Components.WebAssembly.Authentication
+@using {SOLUTION NAMESPACE}.Shared
+@attribute [Authorize]
+@inject HttpClient Http
+
+<h1>Starfleet Starship Database</h1>
+
+<h2>New Ship Entry Form</h2>
+
+<EditForm Model="@starship" OnValidSubmit="HandleValidSubmit">
+    <DataAnnotationsValidator />
+    <ServerValidator @ref="serverValidator" />
+    <ValidationSummary />
+
+    ...
+
+</EditForm>
+
+@code {
+    private ServerValidator serverValidator;
+    private Starship starship = new Starship() { ProductionDate = DateTime.UtcNow };
+
+    private async Task HandleValidSubmit(EditContext editContext)
+    {
+        try
+        {
+            var response = await Http.PostAsJsonAsync<Starship>(
+                "StarshipValidation", (Starship)editContext.Model);
```

Comment:
this is a bad practice : sever side validation should be done when processing the form or else a "hacker" could call the for processing endpoint without the validation. Maybe show here a post to "ProcessStarship" then if you receive a 400 send the error messages to the serverValidator

Reply:
Well .... it's a different scenario tho. Yes, they could do that. All this is trying to show is how to send the data up and come back with errors. It's not making any assumption about where/how the form should be processed. Sure, that's a good idea that you're pitching. It's just not something that this example is designed to do thus far. It's not a "bad practice" if the spec calls for processing the form entirely client-side. We could do either of ...

* As you suggest, assume that the form will be processed by the server API. Take the approach that you suggested.
* Add text to explain that this is only validating aspects of the model for client-side form processing. If the server API is meant to process the form, then additional code is required.

Reply:
You can go for solution 2/ while saying that all validation must be done again before processing data server-side

Reply:
I'm liking the idea of showing this with server-side form validation+processing now in one shot. I think 🤔 your suggestion is solid. :+1:

If a dev only wants client-side processing with only server-side validation, they should be able to downgrade such an example easily enough. The challenging piece is the custom validator component, and that bit wouldn't change for server-side processing.

... and since we're headed that way, we can provide a small **_BONUS!_** I feel like we could disable the form and show a 'form submission successful' message.